File: backend/src/engines/roi_calculator.js
/**
 * ROI Calculator Engine
 * Calculates financial returns on treasury decisions
 * Produces monthly/yearly projections with measurable KPIs
 */

class ROICalculator {
  constructor() {
    this.daysPerYear = 365;
    this.daysPerMonth = 30.44; // Average days per month
  }

  /**
   * Calculate ROI for FX strategy decision
   */
  calculateFXStrategyROI(strategy, baselineAmount, executionCost, savings, timeframeMonths = 1) {
    const timeframeYears = timeframeMonths / 12;
    const savingsPerMonth = savings / timeframeMonths;
    const savingsPerYear = savings / timeframeYears;

    // Calculate ROI percentage
    const roi = (savings / executionCost) * 100;
    const roiAnnualized = roi / timeframeYears;

    // Payback period in days
    const paybackDays = (executionCost / (savings / (timeframeMonths * this.daysPerMonth))).toFixed(1);

    return {
      strategy,
      timeframe: `${timeframeMonths} month${timeframeMonths > 1 ? 's' : ''}`,
      baseline: baselineAmount.toFixed(0),
      executionCost: executionCost.toFixed(0),
      grossSavings: savings.toFixed(0),
      netSavings: (savings - executionCost).toFixed(0),
      
      roi: {
        percentage: roi.toFixed(2) + '%',
        annualized: roiAnnualized.toFixed(2) + '%',
        paybackDays: paybackDays + ' days',
      },

      projections: {
        weekly: (savingsPerMonth / 4.3).toFixed(0),
        monthly: savingsPerMonth.toFixed(0),
        quarterly: (savingsPerMonth * 3).toFixed(0),
        semiAnnual: (savingsPerMonth * 6).toFixed(0),
        annual: savingsPerYear.toFixed(0),
      },

      scorecard: {
        efficiency: roi > 50 ? '⭐⭐⭐⭐⭐' : roi > 20 ? '⭐⭐⭐⭐' : roi > 5 ? '⭐⭐⭐' : '⭐⭐',
        timeToValue: paybackDays < 7 ? 'Immediate' : paybackDays < 30 ? 'Quick (< 30 days)' : 'Medium-term',
        recommendation: roi > 50 ? 'HIGHLY RECOMMENDED' : roi > 20 ? 'RECOMMENDED' : 'CONSIDER',
      },
    };
  }

  /**
   * Calculate ROI for supplier negotiation
   */
  calculateSupplierNegotiationROI(
    supplierName,
    currentPrice,
    negotiatedPrice,
    quantity,
    negotiationEffort = 'BALANCED',
    monthlyRecurrence = true
  ) {
    const savingsPerTransaction = currentPrice - negotiatedPrice;
    const transactionsPerYear = monthlyRecurrence ? 12 : 1;
    const annualSavings = savingsPerTransaction * quantity * transactionsPerYear;

    // Negotiation effort cost (internal resources)
    const effortCosts = {
      'AGGRESSIVE': 5000,
      'BALANCED': 2500,
      'COLLABORATIVE': 1500,
    };
    const effortCost = effortCosts[negotiationEffort] || 2500;

    // First transaction benefit
    const firstTransactionSavings = savingsPerTransaction * quantity;

    // Payback period
    const paybackTransactions = monthlyRecurrence
      ? effortCost / (savingsPerTransaction * quantity)
      : 1;
    const paybackMonths = paybackTransactions * (monthlyRecurrence ? 1 : 12);

    const roi = ((annualSavings - effortCost) / effortCost) * 100;

    return {
      supplier: supplierName,
      currentPrice: currentPrice.toFixed(2),
      negotiatedPrice: negotiatedPrice.toFixed(2),
      discountAmount: savingsPerTransaction.toFixed(2),
      discountPercent: ((savingsPerTransaction / currentPrice) * 100).toFixed(2) + '%',
      quantity,
      
      roi: {
        firstTransaction: firstTransactionSavings.toFixed(0),
        annual: annualSavings.toFixed(0),
        netAnnual: (annualSavings - effortCost).toFixed(0),
        percentage: roi.toFixed(2) + '%',
      },

      effort: {
        estimatedHours: this._effortToHours(negotiationEffort),
        internalCost: effortCost.toFixed(0),
        timeToPayback: paybackMonths.toFixed(1) + ' months',
      },

      projections: {
        monthly: (annualSavings / 12).toFixed(0),
        quarterly: (annualSavings / 4).toFixed(0),
        semiAnnual: (annualSavings / 2).toFixed(0),
        annual: annualSavings.toFixed(0),
        threeYear: (annualSavings * 3).toFixed(0),
        fiveYear: (annualSavings * 5).toFixed(0),
      },

      scorecard: {
        impact: annualSavings > 100000 ? '⭐⭐⭐⭐⭐' : annualSavings > 50000 ? '⭐⭐⭐⭐' : '⭐⭐⭐',
        paybackSpeed: paybackMonths < 1 ? 'Immediate' : paybackMonths < 3 ? 'Very Quick' : paybackMonths < 6 ? 'Quick' : 'Medium-term',
        recommendation: roi > 100 ? 'PRIORITY' : roi > 50 ? 'STRONGLY RECOMMENDED' : roi > 20 ? 'RECOMMENDED' : 'OPTIONAL',
        recurringBenefit: monthlyRecurrence ? 'YES (repeats monthly)' : 'ONE-TIME',
      },
    };
  }

  /**
   * Calculate comprehensive ROI across multiple decisions
   */
  calculateComprehensiveROI(decisions) {
    // decisions = [
    //   { type: 'FX', strategy: 'CONVERT_NOW', savings: 10000, cost: 150, months: 1 },
    //   { type: 'SUPPLIER', supplier: 'SupplierA', annualSavings: 50000, effortCost: 2500 },
    // ]

    let totalSavings = 0;
    let totalCosts = 0;
    const breakdownByType = {};

    decisions.forEach(decision => {
      const savings = decision.savings || decision.annualSavings || 0;
      const cost = decision.cost || decision.effortCost || 0;

      totalSavings += savings;
      totalCosts += cost;

      if (!breakdownByType[decision.type]) {
        breakdownByType[decision.type] = { savings: 0, cost: 0, count: 0 };
      }
      breakdownByType[decision.type].savings += savings;
      breakdownByType[decision.type].cost += cost;
      breakdownByType[decision.type].count += 1;
    });

    const netBenefit = totalSavings - totalCosts;
    const roi = (netBenefit / totalCosts) * 100;
    const roiAnnualized = roi; // Assuming annual basis

    return {
      overview: {
        totalDecisions: decisions.length,
        totalSavings: totalSavings.toFixed(0),
        totalCosts: totalCosts.toFixed(0),
        netBenefit: netBenefit.toFixed(0),
        roi: roi.toFixed(2) + '%',
        roiAnnualized: roiAnnualized.toFixed(2) + '%',
      },

      breakdown: Object.entries(breakdownByType).map(([type, data]) => ({
        type,
        decisions: data.count,
        savingsPerDecision: (data.savings / data.count).toFixed(0),
        totalSavings: data.savings.toFixed(0),
        totalCost: data.cost.toFixed(0),
        netBenefit: (data.savings - data.cost).toFixed(0),
        roi: ((data.savings - data.cost) / data.cost * 100).toFixed(2) + '%',
      })),

      projections: {
        monthly: (totalSavings / 12).toFixed(0),
        quarterly: (totalSavings / 4).toFixed(0),
        annual: totalSavings.toFixed(0),
        threeYear: (totalSavings * 3).toFixed(0),
        fiveYear: (totalSavings * 5).toFixed(0),
      },

      performance: {
        bestDecision: this._identifyBestDecision(decisions),
        quickestPayback: this._quickestPayback(decisions),
        highestImpact: this._highestImpact(decisions),
      },

      scorecard: {
        overallEvaluation: roi > 100 ? '⭐⭐⭐⭐⭐ EXCEPTIONAL' : roi > 50 ? '⭐⭐⭐⭐ EXCELLENT' : roi > 20 ? '⭐⭐⭐ GOOD' : '⭐⭐ MODEST',
        implementationRecommendation: netBenefit > 50000 ? 'IMPLEMENT ALL' : 'IMPLEMENT TOP PRIORITY',
      },
    };
  }

  /**
   * Compare ROI across different scenarios
   */
  compareROIScenarios(scenarioResults) {
    // scenarioResults = [
    //   { scenario: 'Normal', savings: 50000, cost: 2000 },
    //   { scenario: 'Crisis', savings: 120000, cost: 2000 },
    //   { scenario: 'Overpricing', savings: 75000, cost: 3000 },
    // ]

    const analyzed = scenarioResults.map(scenario => {
      const netBenefit = scenario.savings - scenario.cost;
      const roi = (netBenefit / scenario.cost) * 100;

      return {
        ...scenario,
        netBenefit: netBenefit.toFixed(0),
        roi: roi.toFixed(2) + '%',
        recommendation: roi > 50 ? 'RECOMMENDED' : 'ACCEPTABLE',
      };
    });

    const ranked = [...analyzed].sort((a, b) => {
      const aROI = Number(a.roi);
      const bROI = Number(b.roi);
      return bROI - aROI;
    });

    return {
      scenarios: analyzed,
      ranking: ranked.map((s, i) => ({
        rank: i + 1,
        scenario: s.scenario,
        roi: s.roi,
        netBenefit: s.netBenefit,
        recommendation: s.recommendation,
      })),
      bestCase: {
        scenario: ranked[0].scenario,
        roi: ranked[0].roi,
        savingsUpside: (Number(ranked[0].netBenefit) - Number(ranked[1].netBenefit)).toFixed(0),
      },
      worstCase: {
        scenario: ranked[ranked.length - 1].scenario,
        roi: ranked[ranked.length - 1].roi,
        savingsDownside: (Number(ranked[1].netBenefit) - Number(ranked[ranked.length - 1].netBenefit)).toFixed(0),
      },
    };
  }

  /**
   * Create dashboard KPI card
   */
  generateKPICard(metric, value, unit = 'RM', trend = 'neutral', variance = 0) {
    let trendIndicator = '→';
    let trendClass = 'neutral';

    if (trend === 'up' && variance > 0) {
      trendIndicator = '↑';
      trendClass = 'positive';
    } else if (trend === 'down' && variance < 0) {
      trendIndicator = '↓';
      trendClass = 'negative';
    }

    return {
      metric,
      value: typeof value === 'number' ? value.toFixed(0) : value,
      unit,
      trend: trendIndicator,
      trendClass,
      variance: Math.abs(variance).toFixed(0),
      display: `${trendIndicator} ${typeof value === 'number' ? value.toFixed(0) : value} ${unit}`,
    };
  }

  /**
   * Create 90-day projection chart data
   */
  generate90DayProjection(monthlyBenefit, startAmount = 0) {
    const days = [];
    let cumulativeBenefit = startAmount;

    for (let day = 1; day <= 90; day++) {
      cumulativeBenefit += monthlyBenefit / 30.44; // Daily benefit
      days.push({
        day,
        date: new Date(new Date().getTime() + day * 86400000).toISOString().split('T')[0],
        dailyBenefit: (monthlyBenefit / 30.44).toFixed(0),
        cumulativeBenefit: cumulativeBenefit.toFixed(0),
      });
    }

    return {
      period: '90 days',
      monthlyBenefit: monthlyBenefit.toFixed(0),
      projectedTotal: (monthlyBenefit * 3).toFixed(0),
      dayByDayData: days,
      milestones: {
        day30: (monthlyBenefit).toFixed(0),
        day60: (monthlyBenefit * 2).toFixed(0),
        day90: (monthlyBenefit * 3).toFixed(0),
      },
    };
  }

  // Helper methods
  _effortToHours(effort) {
    const hours = {
      'AGGRESSIVE': 20,
      'BALANCED': 12,
      'COLLABORATIVE': 8,
    };
    return hours[effort] || 12;
  }

  _identifyBestDecision(decisions) {
    return decisions.reduce((best, curr) => {
      const currROI = (curr.savings || curr.annualSavings) / (curr.cost || curr.effortCost);
      const bestROI = (best.savings || best.annualSavings) / (best.cost || best.effortCost);
      return currROI > bestROI ? curr : best;
    });
  }

  _quickestPayback(decisions) {
    return decisions.reduce((quickest, curr) => {
      const currPayback = (curr.cost || curr.effortCost) / (curr.savings || curr.annualSavings);
      const quickestPayback = (quickest.cost || quickest.effortCost) / (quickest.savings || quickest.annualSavings);
      return currPayback < quickestPayback ? curr : quickest;
    });
  }

  _highestImpact(decisions) {
    return decisions.reduce((highest, curr) => {
      const currSavings = curr.savings || curr.annualSavings;
      const highestSavings = highest.savings || highest.annualSavings;
      return currSavings > highestSavings ? curr : highest;
    });
  }
}

export default ROICalculator;

File: treasurai-backend/simulation/roi_calculator.py
"""
ROI Calculator Engine
Calculates financial returns on treasury decisions
Produces monthly/yearly projections with measurable KPIs
"""


class ROICalculator:
    def __init__(self):
        self.days_per_year = 365
        self.days_per_month = 30.44  # Average days per month

    def calculate_fx_strategy_roi(self, strategy, baseline_amount, execution_cost, savings, timeframe_months=1):
        """Calculate ROI for FX strategy decision"""
        timeframe_years = timeframe_months / 12
        savings_per_month = savings / timeframe_months
        savings_per_year = savings / timeframe_years

        # Calculate ROI percentage
        roi = (savings / execution_cost) * 100
        roi_annualized = roi / timeframe_years

        # Payback period in days
        payback_days = (execution_cost / (savings / (timeframe_months * self.days_per_month)))

        return {
            'strategy': strategy,
            'timeframe': f'{timeframe_months} month{"s" if timeframe_months > 1 else ""}',
            'baseline': round(baseline_amount, 0),
            'execution_cost': round(execution_cost, 0),
            'gross_savings': round(savings, 0),
            'net_savings': round(savings - execution_cost, 0),
            'roi': {
                'percentage': f'{roi:.2f}%',
                'annualized': f'{roi_annualized:.2f}%',
                'payback_days': f'{payback_days:.1f} days',
            },
            'projections': {
                'weekly': round(savings_per_month / 4.3, 0),
                'monthly': round(savings_per_month, 0),
                'quarterly': round(savings_per_month * 3, 0),
                'semi_annual': round(savings_per_month * 6, 0),
                'annual': round(savings_per_year, 0),
            },
            'scorecard': {
                'efficiency': '⭐⭐⭐⭐⭐' if roi > 50 else '⭐⭐⭐⭐' if roi > 20 else '⭐⭐⭐' if roi > 5 else '⭐⭐',
                'time_to_value': 'Immediate' if payback_days < 7 else 'Quick (< 30 days)' if payback_days < 30 else 'Medium-term',
                'recommendation': 'HIGHLY RECOMMENDED' if roi > 50 else 'RECOMMENDED' if roi > 20 else 'CONSIDER',
            },
        }

    def calculate_supplier_negotiation_roi(
        self,
        supplier_name,
        current_price,
        negotiated_price,
        quantity,
        negotiation_effort='BALANCED',
        monthly_recurrence=True
    ):
        """Calculate ROI for supplier negotiation"""
        savings_per_transaction = current_price - negotiated_price
        transactions_per_year = 12 if monthly_recurrence else 1
        annual_savings = savings_per_transaction * quantity * transactions_per_year

        # Negotiation effort cost (internal resources)
        effort_costs = {
            'AGGRESSIVE': 5000,
            'BALANCED': 2500,
            'COLLABORATIVE': 1500,
        }
        effort_cost = effort_costs.get(negotiation_effort, 2500)

        # First transaction benefit
        first_transaction_savings = savings_per_transaction * quantity

        # Payback period
        payback_transactions = effort_cost / (savings_per_transaction * quantity) if monthly_recurrence else 1
        payback_months = payback_transactions * (1 if monthly_recurrence else 12)

        roi = ((annual_savings - effort_cost) / effort_cost) * 100

        return {
            'supplier': supplier_name,
            'current_price': f'{current_price:.2f}',
            'negotiated_price': f'{negotiated_price:.2f}',
            'discount_amount': f'{savings_per_transaction:.2f}',
            'discount_percent': f'{(savings_per_transaction / current_price) * 100:.2f}%',
            'quantity': quantity,
            'roi': {
                'first_transaction': round(first_transaction_savings, 0),
                'annual': round(annual_savings, 0),
                'net_annual': round(annual_savings - effort_cost, 0),
                'percentage': f'{roi:.2f}%',
            },
            'effort': {
                'estimated_hours': self._effort_to_hours(negotiation_effort),
                'internal_cost': round(effort_cost, 0),
                'time_to_payback': f'{payback_months:.1f} months',
            },
            'projections': {
                'monthly': round(annual_savings / 12, 0),
                'quarterly': round(annual_savings / 4, 0),
                'semi_annual': round(annual_savings / 2, 0),
                'annual': round(annual_savings, 0),
                'three_year': round(annual_savings * 3, 0),
                'five_year': round(annual_savings * 5, 0),
            },
            'scorecard': {
                'impact': '⭐⭐⭐⭐⭐' if annual_savings > 100000 else '⭐⭐⭐⭐' if annual_savings > 50000 else '⭐⭐⭐',
                'payback_speed': 'Immediate' if payback_months < 1 else 'Very Quick' if payback_months < 3 else 'Quick' if payback_months < 6 else 'Medium-term',
                'recommendation': 'PRIORITY' if roi > 100 else 'STRONGLY RECOMMENDED' if roi > 50 else 'RECOMMENDED' if roi > 20 else 'OPTIONAL',
                'recurring_benefit': 'YES (repeats monthly)' if monthly_recurrence else 'ONE-TIME',
            },
        }

    def calculate_comprehensive_roi(self, decisions):
        """Calculate comprehensive ROI across multiple decisions"""
        total_savings = 0
        total_costs = 0
        breakdown_by_type = {}

        for decision in decisions:
            savings = decision.get('savings') or decision.get('annual_savings') or 0
            cost = decision.get('cost') or decision.get('effort_cost') or 0

            total_savings += savings
            total_costs += cost

            decision_type = decision.get('type')
            if decision_type not in breakdown_by_type:
                breakdown_by_type[decision_type] = {'savings': 0, 'cost': 0, 'count': 0}
            breakdown_by_type[decision_type]['savings'] += savings
            breakdown_by_type[decision_type]['cost'] += cost
            breakdown_by_type[decision_type]['count'] += 1

        net_benefit = total_savings - total_costs
        roi = (net_benefit / total_costs) * 100 if total_costs > 0 else 0
        roi_annualized = roi

        return {
            'overview': {
                'total_decisions': len(decisions),
                'total_savings': round(total_savings, 0),
                'total_costs': round(total_costs, 0),
                'net_benefit': round(net_benefit, 0),
                'roi': f'{roi:.2f}%',
                'roi_annualized': f'{roi_annualized:.2f}%',
            },
            'breakdown': [
                {
                    'type': decision_type,
                    'decisions': data['count'],
                    'savings_per_decision': round(data['savings'] / data['count'], 0),
                    'total_savings': round(data['savings'], 0),
                    'total_cost': round(data['cost'], 0),
                    'net_benefit': round(data['savings'] - data['cost'], 0),
                    'roi': f'{((data["savings"] - data["cost"]) / data["cost"] * 100):.2f}%',
                }
                for decision_type, data in breakdown_by_type.items()
            ],
            'projections': {
                'monthly': round(total_savings / 12, 0),
                'quarterly': round(total_savings / 4, 0),
                'annual': round(total_savings, 0),
                'three_year': round(total_savings * 3, 0),
                'five_year': round(total_savings * 5, 0),
            },
            'performance': {
                'best_decision': self._identify_best_decision(decisions),
                'quickest_payback': self._quickest_payback(decisions),
                'highest_impact': self._highest_impact(decisions),
            },
            'scorecard': {
                'overall_evaluation': '⭐⭐⭐⭐⭐ EXCEPTIONAL' if roi > 100 else '⭐⭐⭐⭐ EXCELLENT' if roi > 50 else '⭐⭐⭐ GOOD' if roi > 20 else '⭐⭐ MODEST',
                'implementation_recommendation': 'IMPLEMENT ALL' if net_benefit > 50000 else 'IMPLEMENT TOP PRIORITY',
            },
        }

    def _effort_to_hours(self, effort_level):
        """Convert effort level to estimated hours"""
        effort_map = {
            'AGGRESSIVE': 40,
            'BALANCED': 20,
            'COLLABORATIVE': 10,
        }
        return effort_map.get(effort_level, 20)

    def _identify_best_decision(self, decisions):
        """Identify best decision by ROI"""
        if not decisions:
            return None
        return max(decisions, key=lambda d: d.get('savings', 0))

    def _quickest_payback(self, decisions):
        """Identify decision with quickest payback"""
        if not decisions:
            return None
        return min(decisions, key=lambda d: d.get('cost', float('inf')))

    def _highest_impact(self, decisions):
        """Identify decision with highest impact"""
        if not decisions:
            return None
        return max(decisions, key=lambda d: d.get('savings', 0))
